relay_test.go

```diff
@@ -775,6 +793,80 @@ func TestRelayStalledConnection(t *testing.T) {
 	})
 }
 
+// Test that a stalled connection to the client does not cause stuck calls
+// See https://github.com/uber/tchannel-go/issues/700 for more info.
+func TestRelayStalledClientConnection(t *testing.T) {
+	// This needs to be large enough to fill up the client TCP buffer.
+	const _calls = 100
+
+	opts := testutils.NewOpts().
+		// Expect errors from dropped frames.
+		AddLogFilter("Dropping call due to slow connection to destination.", _calls).
+		AddLogFilter("Couldn't send outbound frame.", _calls).
+		// DisableLogVerification(). // we expect warnings due to removed relay item.
+		SetSendBufferSize(10). // We want to hit the buffer size earlier.
+		SetServiceName("s1").
+		SetRelayOnly()
+	testutils.WithTestServer(t, opts, func(ts *testutils.TestServer) {
+		// Track when the server receives calls
+		gotCall := make(chan struct{}, _calls)
+		testutils.RegisterEcho(ts.Server(), func() {
+			gotCall <- struct{}{}
+		})
+
+		// Create a frame relay that will block all client inbound frames.
+		unblockClientInbound := make(chan struct{})
+		blockerHostPort, relayCancel := testutils.FrameRelay(t, ts.HostPort(), func(outgoing bool, f *Frame) *Frame {
+			if !outgoing && f.Header.ID > 1 {
+				// Block all inbound frames except the initRes
+				<-unblockClientInbound
+			}
+
+			return f
+		})
+		defer relayCancel()
+		defer close(unblockClientInbound)
+
+		client := ts.NewClient(nil)
+
+		ctx, cancel := NewContext(testutils.Timeout(time.Second))
+		defer cancel()
+
+		var calls []*OutboundCall
+
+		// Data to fit one frame fully, large enough to fill TCP buffers.
+		data := bytes.Repeat([]byte("test"), 256*60)
+		for i := 0; i < _calls; i++ {
+			call, err := client.BeginCall(ctx, blockerHostPort, ts.ServiceName(), "echo", nil)
+			require.NoError(t, err, "BeginCall failed")
+
+			require.NoError(t, NewArgWriter(call.Arg2Writer()).Write(nil), "arg2 write failed")
+			require.NoError(t, NewArgWriter(call.Arg3Writer()).Write(data), "arg2 write failed")
+
+			// Wait for server to receive the call
+			<-gotCall
+
+			calls = append(calls, call)
+		}
+
+		// Wait for all calls to end on the relay, and ensure we got failures from the slow client.
+		stats := ts.RelayHost().Stats()
+		stats.WaitForEnd()
+		assert.Contains(t, stats.Map(), "testService-client->s1::echo.failed-relay-dest-conn-slow", "Expect at least 1 failed call due to slow client")
```

Comment:
We're also expecting that none of the calls succeeded, right? Also that the remaining calls should've timed-out?

Reply:
from a relay's perspective, we wrote the response frame, so they "succeeded". we expect them to succeed, but the client won't actually see them due to the block.

unfortunately the write syscall succeeding doesn't actually mean the other side will ever see that data.